src/ILLink.Shared/MessageFormat.cs

```diff
@@ -20,5 +27,19 @@ public static string FormatRequiresAttributeUrlArg (string? url)
 				arg2 = " " + url;
 			return arg2;
 		}
+
+		public static string FormatRequiresAttributeMismatch (bool memberHasAttribute, bool condition, string var0, string var1, string var2)
+		{
+			if (!memberHasAttribute && condition)
+				return string.Format (sharedStrings.GetString ("BaseRequiresMismatchMessage")!, var0, var1, var2);
+			else if (memberHasAttribute && condition)
+				return string.Format (sharedStrings.GetString ("DerivedRequiresMismatchMessage")!, var0, var1, var2);
+			else if (!memberHasAttribute && !condition)
+				return string.Format (sharedStrings.GetString ("InterfaceRequiresMismatchMessage")!, var0, var1, var2);
+			else if (memberHasAttribute && !condition)
+				return string.Format (sharedStrings.GetString ("ImplementationRequiresMismatchMessage")!, var0, var1, var2);
+			else
+				return string.Empty;
```

Comment:
if something should never happen, we should throw an exception to indicate that we don't expect this to be a legal state.